File: gtci/11-tree-breadth-first-search/cpp/6-connect-level-order-siblings.cpp
#include <cassert>
#include <iostream>
#include <queue>

using namespace std;

class TreeNode {
public:
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode *next;

  TreeNode(int x) : val(x), left(nullptr), right(nullptr), next(nullptr) {}

  ~TreeNode() {
    delete left;
    delete right;
  }
};

class Solution {
public:
  TreeNode* connect(TreeNode* root) {
    if (root == nullptr)
      return nullptr;
    
    std::queue<TreeNode*> queue;
    queue.push(root);

    while (!queue.empty()) {
      const size_t levelSize = queue.size();

      for (size_t i = 0; i < levelSize; i++) {
        TreeNode* node = queue.front();
        queue.pop();

        if (i == levelSize - 1 || queue.empty())
          node->next = nullptr;
        else
          node->next = queue.front();
        
        if (node->left != nullptr)
          queue.push(node->left);
        
        if (node->right != nullptr)
          queue.push(node->right);
      }
    }

    return root;
  }
};

int main() {
    Solution sol = Solution();
    
    TreeNode* root1 = new TreeNode(1);
    root1->left = new TreeNode(2);
    root1->right = new TreeNode(3);
    root1->left->left = new TreeNode(4);
    root1->left->right = new TreeNode(5);
    root1->right->left = new TreeNode(6);
    root1->right->right = new TreeNode(7);
    sol.connect(root1);
    assert(root1->next == nullptr);
    assert(root1->left->next == root1->right);
    assert(root1->right->next == nullptr);
    assert(root1->left->left->next == root1->left->right);
    assert(root1->left->right->next == root1->right->left);
    assert(root1->right->left->next == root1->right->right);
    assert(root1->right->right->next == nullptr);
    
    TreeNode* root2 = new TreeNode(12);
    root2->left = new TreeNode(7);
    root2->right = new TreeNode(1);
    root2->left->left = new TreeNode(9);
    root2->right->left = new TreeNode(10);
    root2->right->right = new TreeNode(5);
    sol.connect(root2);
    assert(root2->next == nullptr);
    assert(root2->left->next == root2->right);
    assert(root2->right->next == nullptr);
    assert(root2->left->left->next == root2->right->left);
    assert(root2->right->left->next == root2->right->right);
    assert(root2->right->right->next == nullptr);

    std::cout << "All test cases passed." << std::endl;

    delete root1;
    delete root2;

    return 0;
}

File: gtci/11-tree-breadth-first-search/python3/6-connect-level-order-siblings.py
from collections import deque


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = self.right = self.next = None


class Solution:
    def connect(self, root):
        if root is None:
            return None

        queue = deque()
        queue.append(root)

        while queue:
            level_size = len(queue)

            for i in range(level_size):
                node = queue.popleft()

                if i != level_size - 1:
                    node.next = queue[0]
                else:
                    node.next = None

                if node.left is not None:
                    queue.append(node.left)

                if node.right is not None:
                    queue.append(node.right)

        return root


if __name__ == "__main__":
    sol = Solution()

    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.left.right = TreeNode(5)
    root.right.left = TreeNode(6)
    root.right.right = TreeNode(7)
    sol.connect(root)
    assert root.next is None
    assert root.left.next == root.right
    assert root.right.next is None
    assert root.left.left.next == root.left.right
    assert root.left.right.next == root.right.left
    assert root.right.left.next == root.right.right
    assert root.right.right.next is None

    root = TreeNode(12)
    root.left = TreeNode(7)
    root.right = TreeNode(1)
    root.left.left = TreeNode(9)
    root.right.left = TreeNode(10)
    root.right.right = TreeNode(5)
    sol.connect(root)
    assert root.next is None
    assert root.left.next == root.right
    assert root.right.next is None
    assert root.left.left.next == root.right.left
    assert root.right.left.next == root.right.right
    assert root.right.right.next is None

    print("All test cases passed.")
